src/Microsoft.IdentityModel.Tokens/InMemoryCryptoProviderCache.cs

```diff
@@ -84,23 +84,42 @@ private string GetCacheKeyPrivate(SecurityKey securityKey, string algorithm, str
             return string.Format(CultureInfo.InvariantCulture,
                                  "{0}-{1}-{2}-{3}",
                                  securityKey.GetType(),
-                                 string.IsNullOrEmpty(securityKey.KeyId) ? securityKey.InternalId : securityKey.KeyId,
+                                 securityKey.InternalId,
                                  algorithm,
                                  typeofProvider);
         }
 
+        /// <summary>
+        /// For some security key types, in some runtimes, it's not possible to extract public key material and create <see cref="SecurityKey.InternalId"/>.
+        /// In these cases, <see cref="SecurityKey.InternalId"/> will be an empty string, and these keys should not be cached.
+        /// </summary>
+        /// <param name="signatureProvider"><see cref="SignatureProvider"/> to be examined.</param>
+        /// <returns><c>True</c> if <paramref name="signatureProvider"/> should be cached, <c>false</c> otherwise.</returns>
+        private bool ShouldCacheSignatureProvider(SignatureProvider signatureProvider)
+        {
+            if (string.IsNullOrEmpty(signatureProvider?.Key?.InternalId))
```

Comment:
nit: signatureProvider would not be null for current call graph. Key cannot be null as ctor to SignatureProvider ensures Key is non-null;

Reply:
That's right. I'll remove the null-conditional operators.

Reply:
Addressed with https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1346/commits/0f00581f0a75b4972b983006eda9aad40e728414.